File: java/remoting/main/com/taobao/top/link/remoting/CrossLanguageJsonSerializer.java
package com.taobao.top.link.remoting;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import com.alibaba.fastjson.TypeReference;
import com.alibaba.fastjson.serializer.SerializerFeature;

// design for cross-language
public class CrossLanguageJsonSerializer implements Serializer {
	private static final SerializerFeature[] features = {
			// SerializerFeature.WriteMapNullValue,
			SerializerFeature.WriteNullNumberAsZero,
			SerializerFeature.WriteNullBooleanAsFalse,
			// wrapper.Args = methodCall.Args; maybe raise it
			// SerializerFeature.DisableCircularReferenceDetect
	};

	@Override
	public String getName() {
		return "json";
	}

	@Override
	public byte[] serializeMethodCall(MethodCall methodCall) throws FormatterException {
		MethodCallWrapper wrapper = new MethodCallWrapper(methodCall);
		wrapper.Args = methodCall.Args;
		wrapper.MethodSignature = new String[
				methodCall.MethodSignature != null ? methodCall.MethodSignature.length : 0];
		for (int i = 0; i < methodCall.MethodSignature.length; i++)
			wrapper.MethodSignature[i] = this.parseTypeName(methodCall.MethodSignature[i]);
		return JSON.toJSONBytes(wrapper, features);
	}

	@Override
	public MethodCall deserializeMethodCall(byte[] input) throws FormatterException {
		JSONObject obj = (JSONObject) JSON.parse(input);
		MethodCall methodCall = new MethodCall();
		methodCall.MethodName = obj.getString("MethodName");
		methodCall.TypeName = obj.getString("TypeName");
		methodCall.Uri = obj.getString("Uri");

		JSONArray methodSignature = obj.getJSONArray("MethodSignature");
		if (methodSignature != null) {
			methodCall.MethodSignature = new Class<?>[methodSignature.size()];
			for (int i = 0; i < methodCall.MethodSignature.length; i++) {
				try {
					methodCall.MethodSignature[i] = this.parseType(methodSignature.getString(i));
				} catch (ClassNotFoundException e) {
					throw new FormatterException("parse MethodSignature error", e);
				}
			}
		}

		JSONArray args = obj.getJSONArray("Args");
		if (args != null) {
			methodCall.Args = new Object[args.size()];

			for (int i = 0; i < methodCall.Args.length; i++) {
				// said, java generics was not real type,
				// so HashMap.class did not work for json deserialize,
				// and HashMap<Object, Object>.class was wrong
				if (methodCall.MethodSignature[i].equals(HashMap.class))
					methodCall.Args[i] = JSON.parseObject(
							args.getJSONObject(i).toJSONString(),
							new TypeReference<HashMap<Object, Object>>() {
							});
				else
					methodCall.Args[i] = args.getObject(i, methodCall.MethodSignature[i]);
			}
		}
		return methodCall;
	}

	@Override
	public byte[] serializeMethodReturn(MethodReturn methodReturn) throws FormatterException {
		MethodReturnWrapper wrapper = new MethodReturnWrapper();
		wrapper.ReturnValue = methodReturn.ReturnValue;
		wrapper.Exception = methodReturn.Exception != null ?
				methodReturn.Exception.toString() :
				null;
		return JSON.toJSONBytes(wrapper, features);
	}

	@Override
	public MethodReturn deserializeMethodReturn(byte[] input, Class<?> returnType) throws FormatterException {
		JSONObject obj = (JSONObject) JSON.parse(input);
		MethodReturn methodReturn = new MethodReturn();
		methodReturn.ReturnValue = obj.get("ReturnValue") != null ?
				obj.getObject("ReturnValue", returnType) :
				null;
		// TODO:add error stack support
		String exception = obj.getString("Exception");
		if (exception != null && !exception.equals("")) {
			methodReturn.Exception = new Exception(exception);
			methodReturn.Exception.setStackTrace(new StackTraceElement[0]);
		}
		return methodReturn;
	}

	private String parseTypeName(Class<?> type) {
		if (String.class.equals(type))
			return "";
		if (byte.class.equals(type))
			return "b";
		if (Byte.class.equals(type))
			return "B";
		if (double.class.equals(type))
			return "d";
		if (Double.class.equals(type))
			return "D";
		if (float.class.equals(type))
			return "f";
		if (Float.class.equals(type))
			return "F";
		if (int.class.equals(type))
			return "i";
		if (Integer.class.equals(type))
			return "I";
		if (long.class.equals(type))
			return "l";
		if (Long.class.equals(type))
			return "L";
		if (short.class.equals(type))
			return "s";
		if (Short.class.equals(type))
			return "S";
		if (Date.class.equals(type))
			return "t";
		if (Map.class.equals(type) || Map.class.isAssignableFrom(type))
			return "m";
		if (type.isArray())
			return String.format("[%s", this.parseTypeName(type.getComponentType()));
		return type.getName();
	}

	private Class<?> parseType(String typeName) throws ClassNotFoundException {
		if ("".equalsIgnoreCase(typeName))
			return String.class;
		if ("b".equalsIgnoreCase(typeName))
			return byte.class;
		if ("B".equalsIgnoreCase(typeName))
			return Byte.class;
		if ("d".equalsIgnoreCase(typeName))
			return double.class;
		if ("D".equalsIgnoreCase(typeName))
			return Double.class;
		if ("f".equalsIgnoreCase(typeName))
			return float.class;
		if ("F".equalsIgnoreCase(typeName))
			return Float.class;
		if ("i".equalsIgnoreCase(typeName))
			return int.class;
		if ("I".equalsIgnoreCase(typeName))
			return Integer.class;
		if ("l".equalsIgnoreCase(typeName))
			return long.class;
		if ("L".equalsIgnoreCase(typeName))
			return Long.class;
		if ("s".equalsIgnoreCase(typeName))
			return short.class;
		if ("S".equalsIgnoreCase(typeName))
			return Short.class;
		if ("t".equalsIgnoreCase(typeName))
			return Date.class;
		if ("m".equalsIgnoreCase(typeName))
			return HashMap.class;
		if (typeName.charAt(0) == '[')
			// java array: [Ljava.lang.String
			typeName = String.format("[L%s;",
					this.parseType(this.getComponentTypeName(typeName)).getName());
		return Class.forName(typeName, false, this.getClass().getClassLoader());
	}

	private String getComponentTypeName(String typeName) {
		return typeName.substring(1);
	}
}

File: csharp/src/Top.Link/Remoting/Serialization/Json/CrossLanguageJsonSerializer.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Top.Link.Remoting.Serialization.Json
{
    /// <summary>cross-language remoting serializer, take care of type
    /// </summary>
    public class CrossLanguageJsonSerializer : ISerializer
    {
        private static readonly JsonSerializer jsonSerializer;
        static CrossLanguageJsonSerializer()
        {
            jsonSerializer = JsonSerializer.CreateDefault();
            jsonSerializer.Converters.Add(new DateTimeLongConverter());
        }

        public string Name { get { return "json"; } }
        public byte[] SerializeMethodCall(MethodCall methodCall)
        {
            var wrapper = new MethodCallWrapper(methodCall);
            wrapper.Args = methodCall.Args;
            wrapper.MethodSignature = new string[
                methodCall.MethodSignature != null ? methodCall.MethodSignature.Length : 0];
            for (int i = 0; i < methodCall.MethodSignature.Length; i++)
                wrapper.MethodSignature[i] = this.parseTypeName(methodCall.MethodSignature[i]);
            return this.ToJsonBytes(wrapper);
        }
        public MethodCall DeserializeMethodCall(byte[] input)
        {
            JObject obj = this.ToJsonObject(input);
            var methodCall = new MethodCall();
            methodCall.MethodName = obj.GetValue("MethodName").ToObject<string>();
            methodCall.TypeName = obj.GetValue("TypeName").ToObject<string>();
            methodCall.Uri = obj.GetValue("Uri").ToObject<string>();

            IList<Type> types = new List<Type>();
            foreach (JToken token in obj.GetValue("MethodSignature").AsJEnumerable())
                types.Add(this.parseType(token.ToObject<string>()));
            methodCall.MethodSignature = new Type[types.Count];
            types.CopyTo(methodCall.MethodSignature, 0);

            var i = 0;
            IList<object> args = new List<object>();
            foreach (JToken token in obj.GetValue("Args").AsJEnumerable())
                args.Add(token.ToObject(methodCall.MethodSignature[i++], jsonSerializer));
            methodCall.Args = new object[args.Count];
            args.CopyTo(methodCall.Args, 0);

            return methodCall;
        }
        public byte[] SerializeMethodReturn(MethodReturn methodReturn)
        {
            var wrapper = new MethodReturnWrapper();
            wrapper.ReturnValue = methodReturn.ReturnValue;
            wrapper.Exception = methodReturn.Exception != null ?
                    methodReturn.Exception.Message :
                    null;
            return this.ToJsonBytes(wrapper);
        }
        public MethodReturn DeserializeMethodReturn(byte[] input, Type returnType)
        {
            JToken t;
            JObject obj = this.ToJsonObject(input);
            MethodReturn methodReturn = new MethodReturn();

            if (obj.TryGetValue("ReturnValue", out t))
                methodReturn.ReturnValue = t.ToObject(returnType, jsonSerializer);

            var ex = obj.TryGetValue("Exception", out t) ? t.ToObject<string>() : null;
            if (!string.IsNullOrEmpty(ex))
                methodReturn.Exception = new Exception(ex);

            return methodReturn;
        }

        private string parseTypeName(Type type)
        {
            if (typeof(string).Equals(type))
                return "";
            if (typeof(byte).Equals(type))
                return "b";
            if (typeof(Byte).Equals(type))
                return "B";
            if (typeof(double).Equals(type))
                return "d";
            if (typeof(Double).Equals(type))
                return "D";
            if (typeof(float).Equals(type))
                return "f";
            if (typeof(int).Equals(type))
                return "i";
            if (typeof(long).Equals(type))
                return "l";
            if (typeof(short).Equals(type))
                return "s";
            if (typeof(DateTime).Equals(type))
                return "t";
            if (typeof(IDictionary<string, string>).Equals(type)
                || typeof(IDictionary<string, object>).Equals(type)
                || typeof(IDictionary<object, object>).Equals(type)
                || typeof(IDictionary<string, string>).IsAssignableFrom(type)
                || typeof(IDictionary<string, object>).IsAssignableFrom(type)
                || typeof(IDictionary<object, object>).IsAssignableFrom(type))
                return "m";
            if (type.IsArray)
                return string.Format("[{0}", this.parseTypeName(type.GetElementType()));
            return type.FullName;
        }
        private Type parseType(string typeName)
        {
            if ("".Equals(typeName))
                return typeof(string);
            if ("b".Equals(typeName))
                return typeof(byte);
            if ("B".Equals(typeName))
                return typeof(Byte);
            if ("d".Equals(typeName))
                return typeof(double);
            if ("D".Equals(typeName))
                return typeof(Double);
            if ("f".Equals(typeName))
                return typeof(float);
            if ("i".Equals(typeName))
                return typeof(int);
            if ("l".Equals(typeName))
                return typeof(long);
            if ("s".Equals(typeName))
                return typeof(short);
            if ("t".Equals(typeName))
                return typeof(DateTime);
            if ("m".Equals(typeName))
                return typeof(IDictionary<string, string>);
            if (typeName[0] == '[')
                // c# array: System.String[]
                typeName = string.Format("{0}[]", this.parseType(this.GetElementTypeName(typeName)).FullName);
            //TODO:support special assembly for searching
            foreach (var a in AppDomain.CurrentDomain.GetAssemblies())
            {
                var t = a.GetType(typeName, false);
                if (t != null)
                    return t;
            }
            return Type.GetType(typeName, true);
        }
        private byte[] ToJsonBytes(object value)
        {
            //instead of JsonConvert.SerializeObject
            //https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/JsonConvert.cs#L588
            //using (var s = new MemoryStream())
            //using (var w = new StreamWriter(s, Encoding.UTF8))
            //MemoryStream not work?
            var s = new StringBuilder();
            var w = new StringWriter(s);
            using (var jsonWriter = new JsonTextWriter(w))
            {
                jsonWriter.Formatting = Formatting.None;
                jsonSerializer.Serialize(jsonWriter, value);
                //Console.WriteLine(s.ToString());
                return Encoding.UTF8.GetBytes(s.ToString());
                //return s.ToArray();
            }
        }
        private JObject ToJsonObject(byte[] input)
        {
            //Console.WriteLine(Encoding.UTF8.GetString(input));
            //instead of JsonConvert.DeserializeObject
            //https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/JsonConvert.cs#L786
            using (var s = new MemoryStream(input))
            using (var r = new StreamReader(s, Encoding.UTF8))
                return jsonSerializer.Deserialize(new JsonTextReader(r), null) as JObject;
        }
        private string GetElementTypeName(string typeName)
        {
            return typeName.Substring(1);
        }
        public class MethodCallWrapper : MethodCall
        {
            public new String[] MethodSignature { get; set; }
            public MethodCallWrapper() { }
            public MethodCallWrapper(MethodCall origin)
            {
                this.MethodName = origin.MethodName;
                this.TypeName = origin.TypeName;
                this.Uri = origin.Uri;
            }
        }
        public class MethodReturnWrapper : MethodReturn
        {
            public new string Exception { get; set; }
        }
        /// <summary>DateTime/Long Ticks Converter
        /// https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Converters/DateTimeConverterBase.cs
        /// </summary>
        class DateTimeLongConverter : Newtonsoft.Json.Converters.DateTimeConverterBase
        {
            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
            {
                if (reader.TokenType != JsonToken.Integer)
                    throw new Exception(string.Format("Unexpected token parsing date. Expected Integer, got {0}.", reader.TokenType));
                return new DateTime((long)reader.Value);
            }
            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
            {
                if (!(value is DateTime))
                    throw new Exception("Expected date object value.");
                writer.WriteValue(((DateTime)value).Ticks);
            }
        }
    }
}